File: vh_racing/src/components/Car.js
class MassCategory {
    static Light = 'Light';
    static Medium = 'Medium';
    static Heavy = 'Heavy';
}

class TireType {
    static Rain = 'Rain';
    static Slick = 'Slick';
}

class SurfaceType {
    static Asphalt = 'Asphalt';
    static Gravel = 'Gravel';
    static Ice = 'Ice';
}

/**
 * @definition Car
 * @params mCat, tType, startX, startY
 * @returns Car object
 * Car class representing a vehicle with properties for mass category, tire type, position, and various physics-related attributes.
 */
class Car {
    constructor(mCat, tType, startX, startY) {
        this.startX = startX;
        this.startY = startY;
        this.positionX = startX;
        this.positionY = startY;
        this.massCategory = mCat;
        this.tireType = tType;
        this.surfaceType = SurfaceType.Asphalt;
        this.angle = 0.0;
        this.velocity = 0.0;
        this.acceleration = 0.0;
        this.tireTemperature = 20.0;
        this.tireGrip = null;
        this.baseTireGrip = null;
        this.minTireGrip = 0.5;
        this.gripDecayRate = 0.001;
        this.traction = 1.0;

        this.setMassCategory(mCat);
        this.setTireType(tType);

        if (mCat === MassCategory.Light && tType === TireType.Slick) {
            this.maxVelocity = 500.0;
        } else if (mCat === MassCategory.Medium && tType === TireType.Slick) {
            this.maxVelocity = 480.0;
        } else if (mCat === MassCategory.Heavy && tType === TireType.Slick) {
            this.maxVelocity = 460.0;
        } else if (tType === TireType.Rain) {
            this.maxVelocity = 440.0;
        } else {
            this.maxVelocity = 480.0;
        }

        this.wheelbase = 2.5;
        this.setCorneringStiffness();

        this.steeringAngle = 0.0;
        this.maxSteeringAngle = this.degToRad(45);
        this.steeringSpeed = this.degToRad(60);
    }

    /**
     * @definition setMassCategory
     * @params mCat
     * @returns None
     * Sets the mass of the car based on its mass category.
     */
    setMassCategory(mCat) {
        if (mCat === MassCategory.Light) {
            this.mass = 1800.0 * 0.453592;
        } else if (mCat === MassCategory.Medium) {
            this.mass = 2800.0 * 0.453592;
        } else if (mCat === MassCategory.Heavy) {
            this.mass = 3800.0 * 0.453592;
        }
    }

    /**
     * @definition setTireType
     * @params tType
     * @returns None
     * Sets the base tire grip based on the tire type.
     */
    setTireType(tType) {
        if (tType === TireType.Rain) {
            this.baseTireGrip = 0.7;
        } else if (tType === TireType.Slick) {
            this.baseTireGrip = 1.0;
        }
        this.tireGrip = this.baseTireGrip;
    }

    /**
     * @definition setCorneringStiffness
     * @params None
     * @returns None
     * Sets the cornering stiffness values for the front and rear tires based on tire type.
     */
    setCorneringStiffness() {
        let baseStiffness;
        if (this.tireType === TireType.Slick) {
            baseStiffness = 150000;
        } else if (this.tireType === TireType.Rain) {
            baseStiffness = 100000;
        }

        this.Cf = baseStiffness * 0.6;
        this.Cr = baseStiffness * 0.4;
    }

    /**
     * @definition calculateUndersteerGradient
     * @params None
     * @returns Number
     * Calculates and returns the understeer gradient based on car weight distribution and cornering stiffness.
     */
    calculateUndersteerGradient() {
        const Wf = this.mass * 0.6;
        const Wr = this.mass * 0.4;
        const K = (this.wheelbase / this.mass) * ((Wf / this.Cf) - (Wr / this.Cr));
        return K;
    }

    /**
     * @definition applyThrottle
     * @params throttle, deltaTime
     * @returns None
     * Applies throttle to the car, increasing its velocity.
     */
    applyThrottle(throttle, deltaTime) {
        const maxAcceleration = 90.0;
        let acceleration = throttle * maxAcceleration * this.traction * (1 - this.velocity / this.maxVelocity);
        if (acceleration < 0) acceleration = 0;
        this.acceleration = acceleration;
        this.velocity += this.acceleration * deltaTime;
        if (this.velocity > this.maxVelocity) this.velocity = this.maxVelocity;
    }

    /**
     * @definition applyBrake
     * @params brakeForce, deltaTime
     * @returns None
     * Applies brakes to the car, reducing its velocity.
     */
    applyBrake(brakeForce, deltaTime) {
        const maxDeceleration = 60.0;
        const deceleration = brakeForce * maxDeceleration * this.traction;
        this.acceleration = -deceleration;
        this.velocity += this.acceleration * deltaTime;
        if (this.velocity < 0) {
            this.velocity = 0.0;
            this.acceleration = 0.0;
        }
    }

    /**
     * @definition updateSteering
     * @params steeringInput, deltaTime
     * @returns None
     * Updates the steering angle of the car based on steering input.
     */
    updateSteering(steeringInput, deltaTime) {
        this.steeringAngle += steeringInput * this.steeringSpeed * deltaTime;
        if (this.steeringAngle > this.maxSteeringAngle) {
            this.steeringAngle = this.maxSteeringAngle;
        } else if (this.steeringAngle < -this.maxSteeringAngle) {
            this.steeringAngle = -this.maxSteeringAngle;
        }
    }

    /**
     * @definition updatePosition
     * @params deltaTime
     * @returns None
     * Updates the car's position based on its velocity and steering angle.
     */
    updatePosition(deltaTime) {
        const rollingResistance = 12.0;
        const dragCoefficient = 0.4257;
        const airDensity = 1.225;
        const frontalArea = 2.2;
        const dragForce = 0.5 * dragCoefficient * airDensity * frontalArea * this.velocity ** 2;

        const totalResistance = (rollingResistance + dragForce) / this.mass;
        this.velocity -= totalResistance * deltaTime;
        if (this.velocity < 0) this.velocity = 0;

        if (this.velocity > 0) {
            const K = this.calculateUndersteerGradient();
            let adjustedSteeringAngle = this.steeringAngle * (1 + K);
            const speedFactor = Math.max(0.5, 1 - (this.velocity / this.maxVelocity) * 0.5);
            const steeringEffectiveness = 0.9 * speedFactor;
            adjustedSteeringAngle *= steeringEffectiveness;

            this.updateTireGrip();
            this.traction = this.tireGrip;

            const turningRadius = adjustedSteeringAngle !== 0 ? this.wheelbase / Math.tan(adjustedSteeringAngle) : Infinity;
            const angularVelocity = this.velocity / turningRadius;
            this.angle += angularVelocity * deltaTime;
            this.angle = this.angle % (2 * Math.PI);

            const forwardX = this.velocity * Math.cos(this.angle) * deltaTime;
            const forwardY = this.velocity * Math.sin(this.angle) * deltaTime;
            this.positionX += forwardX;
            this.positionY += forwardY;

            this.updateTireTemperature(deltaTime);
            this.applyTireStress(this.steeringAngle, this.velocity, deltaTime);
        }
    }

    /**
     * @definition updateTireTemperature
     * @params deltaTime
     * @returns None
     * Updates the tire temperature based on steering angle and velocity.
     */
    updateTireTemperature(deltaTime) {
        const tempIncrease = (Math.abs(this.steeringAngle) + 0.1) * this.velocity * 0.05 * deltaTime;
        const tempDecrease = (this.tireTemperature - 20.0) * 0.1 * deltaTime;
        this.tireTemperature += tempIncrease - tempDecrease;
        if (this.tireTemperature < 20.0) {
            this.tireTemperature = 20.0;
        }
    }

    /**
     * @definition updateTireGrip
     * @params None
     * @returns None
     * Updates the tire grip based on temperature, speed, and steering angle.
     */
    updateTireGrip() {
        const optimalTemp = 90.0;
        const tempDifference = Math.abs(this.tireTemperature - optimalTemp);
        const temperatureEffect = Math.max(0.5, 1 - (tempDifference / 100));

        const speedEffect = Math.max(0.7, 1 - (this.velocity / this.maxVelocity) * 0.3);
        const steeringEffect = Math.max(0.7, 1 - (Math.abs(this.steeringAngle) / this.maxSteeringAngle) * 0.3);

        this.tireGrip = this.baseTireGrip * temperatureEffect * speedEffect * steeringEffect;
        if (this.tireGrip < this.minTireGrip) {
            this.tireGrip = this.minTireGrip;
        }
    }

    /**
     * @definition applyTireStress
     * @params steeringAngle, velocity, deltaTime
     * @returns None
     * Applies stress to the tires based on steering angle and velocity, decaying the tire grip.
     */
    applyTireStress(steeringAngle, velocity, deltaTime) {
        const stressFactor = (Math.abs(steeringAngle) + 0.1) * velocity;
        const decayAmount = stressFactor * this.gripDecayRate * deltaTime;
        this.tireGrip -= decayAmount;
        if (this.tireGrip < this.minTireGrip) {
            this.tireGrip = this.minTireGrip;
        }
    }

    /**
     * @definition resetState
     * @params None
     * @returns None
     * Resets the car to its initial starting position and orientation.
     */
    resetState() {
        this.positionX = this.startX;
        this.positionY = this.startY;
        this.velocity = 0.0;
        this.acceleration = 0.0;
        this.steeringAngle = 0.0;
        this.angle = 0.0;
    }

    /**
     * @definition degToRad
     * @params deg
     * @returns Number
     * Converts degrees to radians.
     */
    degToRad(deg) {
        return deg * (Math.PI / 180);
    }

    /**
     * @definition radToDeg
     * @params rad
     * @returns Number
     * Converts radians to degrees.
     */
    radToDeg(rad) {
        return rad * (180 / Math.PI);
    }

    /**
     * @definition getPositionX
     * @params None
     * @returns Number
     * Returns the car's X position.
     */
    getPositionX() {
        return this.positionX;
    }

    /**
     * @definition getPositionY
     * @params None
     * @returns Number
     * Returns the car's Y position.
     */
    getPositionY() {
        return this.positionY;
    }

    /**
     * @definition getVelocity
     * @params None
     * @returns Number
     * Returns the car's velocity.
     */
    getVelocity() {
        return this.velocity;
    }

    /**
     * @definition getAcceleration
     * @params None
     * @returns Number
     * Returns the car's acceleration.
     */
    getAcceleration() {
        return this.acceleration;
    }

    /**
     * @definition getTireGrip
     * @params None
     * @returns Number
     * Returns the car's tire grip.
     */
    getTireGrip() {
        return this.tireGrip;
    }

    /**
     * @definition getTireTemperature
     * @params None
     * @returns Number
     * Returns the car's tire temperature.
     */
    getTireTemperature() {
        return this.tireTemperature;
    }

    /**
     * @definition getSteeringAngleDegrees
     * @params None
     * @returns Number
     * Returns the steering angle in degrees.
     */
    getSteeringAngleDegrees() {
        return this.radToDeg(this.steeringAngle);
    }
}

export { Car, MassCategory, TireType, SurfaceType };

File: src/car.py
from enum import Enum
import math

## LEGACY CODE ##

class MassCategory(Enum):
    Light = 0
    Medium = 1
    Heavy = 2

class TireType(Enum):
    Rain = 0
    Slick = 1

class SurfaceType(Enum):
    Asphalt = 0
    Gravel = 1
    Ice = 2

class Car:
    def __init__(self, mCat, tType, startX, startY):
        self.positionX = startX
        self.positionY = startY
        self.massCategory = mCat
        self.tireType = tType
        self.surfaceType = SurfaceType.Asphalt
        self.angle = 0.0
        self.velocity = 0.0
        self.acceleration = 0.0
        self.tireTemperature = 20.0
        self.tireGrip = None
        self.base_tireGrip = None
        self.minTireGrip = 0.5
        self.gripDecayRate = 0.001
        self.traction = 1.0
        self.setMassCategory(mCat)
        self.setTireType(tType)

        if mCat == MassCategory.Light and tType == TireType.Slick:
            self.maxVelocity = 200.0
        elif mCat == MassCategory.Medium and tType == TireType.Slick:
            self.maxVelocity = 180.0
        elif mCat == MassCategory.Heavy and tType == TireType.Slick:
            self.maxVelocity = 160.0
        elif tType == TireType.Rain:
            self.maxVelocity = 140.0
        else:
            self.maxVelocity = 180.0

        self.wheelbase = 2.5
        self.setCorneringStiffness()

        self.steering_angle = 0.0
        self.max_steering_angle = math.radians(45)
        self.steering_speed = math.radians(60)

    def setMassCategory(self, mCat):
        if mCat == MassCategory.Light:
            self.mass = 1800.0 * 0.453592
        elif mCat == MassCategory.Medium:
            self.mass = 2800.0 * 0.453592
        elif mCat == MassCategory.Heavy:
            self.mass = 3800.0 * 0.453592

    def setTireType(self, tType):
        if tType == TireType.Rain:
            self.base_tireGrip = 0.7
        elif tType == TireType.Slick:
            self.base_tireGrip = 1.0
        self.tireGrip = self.base_tireGrip

    def setCorneringStiffness(self):
        if self.tireType == TireType.Slick:
            base_stiffness = 150000
        elif self.tireType == TireType.Rain:
            base_stiffness = 100000

        self.Cf = base_stiffness * 0.6
        self.Cr = base_stiffness * 0.4

    def calculateUndersteerGradient(self):
        Wf = self.mass * 0.6
        Wr = self.mass * 0.4
        K = (self.wheelbase / self.mass) * ((Wf / self.Cf) - (Wr / self.Cr))
        return K

    def applyThrottle(self, throttle, deltaTime):
        maxAcceleration = 30.0
        acceleration = throttle * maxAcceleration * self.traction * (1 - self.velocity / self.maxVelocity)
        if acceleration < 0:
            acceleration = 0
        self.acceleration = acceleration
        self.velocity += self.acceleration * deltaTime
        if self.velocity > self.maxVelocity:
            self.velocity = self.maxVelocity

        # Debug statement for throttle application
        print(
            f"Throttle: {throttle}, Acceleration: {self.acceleration}, Velocity: {self.velocity}, Position: ({self.positionX}, {self.positionY})")

    def applyBrake(self, brakeForce, deltaTime):
        maxDeceleration = 50.0
        deceleration = brakeForce * maxDeceleration * self.traction
        self.acceleration = -deceleration
        self.velocity += self.acceleration * deltaTime
        if self.velocity < 0:
            self.velocity = 0.0
            self.acceleration = 0.0

        # Debug statement for brake application
        print(
            f"Brake: {brakeForce}, Deceleration: {self.acceleration}, Velocity: {self.velocity}, Position: ({self.positionX}, {self.positionY})")

    def updateSteering(self, steeringInput, deltaTime):
        self.steering_angle += steeringInput * self.steering_speed * deltaTime
        if self.steering_angle > self.max_steering_angle:
            self.steering_angle = self.max_steering_angle
        elif self.steering_angle < -self.max_steering_angle:
            self.steering_angle = -self.max_steering_angle

        # Debug statement for steering input
        print(f"Steering input: {steeringInput}, Steering angle (deg): {math.degrees(self.steering_angle)}")

    def updatePosition(self, deltaTime):
        rollingResistance = 12.0
        dragCoefficient = 0.4257
        airDensity = 1.225
        frontalArea = 2.2
        dragForce = 0.5 * dragCoefficient * airDensity * frontalArea * self.velocity ** 2

        totalResistance = (rollingResistance + dragForce) / self.mass
        self.velocity -= totalResistance * deltaTime
        if self.velocity < 0:
            self.velocity = 0

        if self.velocity > 0:
            if self.steering_angle == 0.0:
                self.steering_angle = 0.01
            K = self.calculateUndersteerGradient()
            adjustedSteeringAngle = self.steering_angle * (1 + K)

            speed_factor = max(0.5, 1 - (self.velocity / self.maxVelocity) * 0.5)
            steering_effectiveness = 0.9 * speed_factor
            adjustedSteeringAngle *= steering_effectiveness

            self.updateTireGrip()
            self.traction = self.tireGrip

            if adjustedSteeringAngle != 0:
                turning_radius = self.wheelbase / math.tan(adjustedSteeringAngle)
            else:
                turning_radius = float('inf')

            angular_velocity = self.velocity / turning_radius if turning_radius != float('inf') else 0.0
            self.angle += angular_velocity * deltaTime
            self.angle = self.angle % (2 * math.pi)

            forwardX = self.velocity * math.cos(self.angle) * deltaTime
            forwardY = self.velocity * math.sin(self.angle) * deltaTime
            self.positionX += forwardX
            self.positionY += forwardY

            self.updateTireTemperature(deltaTime)
            self.applyTireStress(self.steering_angle, self.velocity, deltaTime)

        # Debug statement for position update
        print(
            f"Updated position: ({self.positionX}, {self.positionY}), Velocity: {self.velocity}, Steering angle: {math.degrees(self.steering_angle)}")

        if abs(self.steering_angle) > 0:
            steeringReturnSpeed = math.radians(30)
            steeringReturn = steeringReturnSpeed * deltaTime
            if self.steering_angle > 0:
                self.steering_angle -= min(steeringReturn, self.steering_angle)
            else:
                self.steering_angle += min(steeringReturn, -self.steering_angle)

    def updateTireTemperature(self, deltaTime):
        tempIncrease = (abs(self.steering_angle) + 0.1) * self.velocity * 0.05 * deltaTime
        tempDecrease = (self.tireTemperature - 20.0) * 0.1 * deltaTime
        self.tireTemperature += tempIncrease - tempDecrease
        if self.tireTemperature < 20.0:
            self.tireTemperature = 20.0

    def updateTireGrip(self):
        optimalTemp = 90.0
        tempDifference = abs(self.tireTemperature - optimalTemp)
        temperatureEffect = max(0.5, 1 - (tempDifference / 100))

        speedEffect = max(0.7, 1 - (self.velocity / self.maxVelocity) * 0.3)
        steeringEffect = max(0.7, 1 - ((abs(self.steering_angle) + 0.1) / self.max_steering_angle) * 0.3)

        self.tireGrip = self.base_tireGrip * temperatureEffect * speedEffect * steeringEffect
        if self.tireGrip < self.minTireGrip:
            self.tireGrip = self.minTireGrip

        # Debug statement for tire grip
        print(f"Tire temperature: {self.tireTemperature}, Tire grip: {self.tireGrip}")

    def applyTireStress(self, steeringAngle, velocity, deltaTime):
        stressFactor = (abs(steeringAngle) + 0.1) * velocity
        decayAmount = stressFactor * self.gripDecayRate * deltaTime
        self.tireGrip -= decayAmount
        if self.tireGrip < self.minTireGrip:
            self.tireGrip = self.minTireGrip

    def getPositionX(self):
        return self.positionX

    def getPositionY(self):
        return self.positionY

    def getVelocity(self):
        return self.velocity

    def getAcceleration(self):
        return self.acceleration

    def getTireGrip(self):
        return self.tireGrip

    def getTireTemperature(self):
        return self.tireTemperature

    def getSteeringAngleDegrees(self):
        return math.degrees(self.steering_angle)
